lib/clients/websocket.js

```diff
@@ -48,28 +48,68 @@ class WebsocketClient extends EventEmitter {
     this.socket = null;
   }
 
-  onOpen() {
-    this.emit('open');
-
+  sendSubscription(type, productIDs, channels) {
     const subscribeMessage = {
-      type: 'subscribe',
-      product_ids: this.productIDs,
-      channels: this.channels,
+      type: type,
+      channels: [...channels],
     };
 
+    if (productIDs) {
+      subscribeMessage.product_ids = [...productIDs];
+    }
+
     // Add Signature
     if (this.auth.secret) {
-      let sig = signRequest(
-        this.auth,
-        'GET',
-        this.channels ? '/users/self/verify' : '/users/self'
-      );
+      let sig = signRequest(this.auth, 'GET', '/users/self/verify');
       Object.assign(subscribeMessage, sig);
     }
 
     this.socket.send(JSON.stringify(subscribeMessage));
   }
 
+  subscribeProduct(...productIDs) {
+    for (let p of productIDs) {
+      this.productIDs.add(p);
+    }
+
+    this.sendSubscription('subscribe', this.productIDs, this.channels);
+  }
+
+  subscribeChannel(...channels) {
+    for (let c of channels) {
+      this.channels.add(c);
+    }
+
+    this.sendSubscription('subscribe', this.productIDs, this.channels);
+  }
+
+  unsubscribeProduct(...productIDs) {
+    for (let p of productIDs) {
+      this.productIDs.delete(p);
+    }
+
+    var channels = [...this.channels].map(c => ({
+      name: c,
+      product_ids: productIDs,
+    }));
+
+    this.sendSubscription('unsubscribe', null, channels);
+  }
+
+  unsubscribeChannel(...channels) {
+    for (let c of channels) {
+      this.channels.delete(c);
+    }
+
+    this.sendSubscription('unsubscribe', null, channels);
+  }
+
+  onOpen() {
+    this.emit('open');
+
+    this.sendSubscription('subscribe', this.productIDs, this.channels);
```

Comment:
Perhaps, `sendSubscription()` should default the last two arguments to `this.productIDs` and `this.channels`. i.e. Passing around properties (`this.productIDs` and `this.channels`) as arguments to instance methods that already have access to these properties is bit of a OO smell.

In other words, everytime you want to send a `subscribe` message to re-sync the products and channels, you can just call `this.sendSubscription('subscribe')`.

`sendSubscription()` should probably also be made _"privat"e_ and renamed to `_sendSubscription()`.

Reply:
I actually had those default arguments, but I took them out because it does create the misleading perception that `sendSubscription('subscribe')` might set the server channels/products to the class ones, when in reality it will only ensure the class ones are subscribed to, but not unsubscribe from any that have been removed.  They must be passed with 'unsubscribe' for that.  But I'll put 'em back!